### DISCUSSION TITLE
Best practices for efficient memory usage (input data, cube structure)

### DISCUSSION THREAD

**Discussion by lguigo** [atoti/atoti#525] [General]
Hello,

In the existing threads and documentation I couldn't find guidance on how to use memory efficiently. Although very helpful, the JVM Memory reports logged every 30sec do not indicate its breakdown.

Could you share general guidelines on that topic (e.g. hierarchies and measures types to avoid/prefer), and if possible an idea of how memory is generally broken down for a cube?

Thanks,
Louis

**Comment by Acriss** [comment]
Hi Iguigo !

That's a good (and complex) question!
There are two components to this question:

- Data storage
- Transient memory usage at query time

Full disclosure, Atoti relies on [ActivePivot](https://activeviam.com/activepivot/5.11.1/docs/intro/overview/) behind the scene. I'm linking its documentation if you ever want to dive a bit more in depth, and will try to give a general overview below.

When you load data into atoti, it is stored in the Datastore component. You can expect some data compression, indexing, etc..., and can expect this to improve as atoti versions release. It is an in-memory columnar database.

The cube is built on top of this database. The selected columns become hierarchies, but that shouldn't be a memory issue unless you force a very high cardinality column to become a hierarchy (typically, an `id` column where all elements are differents).

The cube reasons in terms of aggregates: the value of a measure at a specific location in the cube. (Think, the value of the measure "number of sales" at the location "city = London, seller = M&S, date = yesterday").
These aggregates can be either stored or calculated on the fly (by querying the datastore). The perfect configuration for a given project will be a mix of the two.

When they are stored, we talk about a component called "aggregates provider". This is, at least for now, a hidden functionality that is automatically configured.

As the measure count grows, this second form of storage will also grow, which is one of the reason we mostly depend on the "on the fly calculation" pattern.


At query time, atoti requires some memory left for computation. We usually recommend having the storage part represent between a third and half of the memory allocated to atoti. The rest will be used by the application throughout its runtime.

There aren't really any hierarchy type to avoid/prefer. However, when you use the `session.tables.schema` visualization, you might see some dotted lines representing relationships between tables. These are partial relationships, meaning that not all key fields are linked to the store on the other side.
Measures across these partial relationships are the most costly, both in terms of CPU and memory usage, as the resolution of the join and the aggregation along this join are a complicated process.

If you need to optimize your application, these are probably the one factor to take care about.

Now I'll leave this overview as is, and feel free to ask if that brings new/more precise questions on your side!
Cheers,
Acriss

**Reply by Acriss** [reply]
Knowing that behind the scene, this is a Java Library, you can use [Java Mission Control](https://www.oracle.com/java/technologies/javase/products-jmc8-downloads.html) to inspect the exposed MBeans, which will provide a lot more details regarding your project.

**Reply by lguigo** [reply]
Hi Acriss, thanks a lot for your detailed answer and apologies for the delay of mine!

We tried a couple of things to assess the memory impacts of various aspects, focusing on the data storage part (rather than the variations due to usage/query over the lifetime of a cube, which seems very stable for our usage so far).

We see a relatively important factor between the size of the input data (pandas dataframe) versus the size of the used heap memory of the Java process containing the cube (up to 10x).

We attempted to profile this live Java process using JMC but it seems like the recommended way to use it is not compatible with Atoti v0.6.4's Java version / openjdk 14.0.2:

            Unrecognized VM option 'UnlockCommercialFeatures'
            Error: Could not create the Java Virtual Machine.
            Error: A fatal exception has occurred. Program will exit.

We will look into different techniques to profile (e.g. netbeans), but could you explain which version of JMC you use / how you use it, on an example perhaps?

Will keep you posted on any progress we make, thanks again for your help.

Louis

**Reply by Acriss** [reply]
Sounds good !

Regarding the analysis of the memory, we do have [a specific tool](https://github.com/activeviam/mac), but it's not yet integrated with atoti. That should come later in the year, as it's on our roadmap.

> We see a relatively important factor between the size of the input data (pandas dataframe) versus the size of the used heap memory of the Java process containing the cube (up to 10x).

Could you share what's the size of your input dataframe ?

Storage in atoti is done Off-Heap rather than On-Heap, to alleviate the pressure on the Garbage Collector, and to allow us to associate data to a particular [Numa node](https://activeviam.com/activepivot/5.11.1/docs/concepts/partitioning/#numa-awareness-and-optimization) .
You can vaguely monitor it using the JConsole tool.

The OnHeap memory is taken by everything that is not raw data in atoti.

> We attempted to profile this live Java process using JMC but it seems like the recommended way to use it is not compatible with atoti v0.6.4's Java version / openjdk 14.0.2

Starting with Java 11 (or maybe even 9), the Java Flight Recorder is open source, and does not require the `UnlockCommercialFeatures` when using OpenJDK .

Please do keep us posted !
Acris
